test/models/User.spec.ts

```diff
@@ -180,4 +178,94 @@ describe("User", () => {
+    it("should throw a NoCodeError if a user tries to validate a code but no code is stored", async () => {
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "authentication-code":
+              return null;
+          }
+        },
+      });
+
+      try {
+        await caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' });
+        throw new Error('Should throw')
+      } catch (err: any) {
+        if (err.message !== 'No code found') {
+          throw new Error('Should throw NoCodeError')
+        }
+      }
+    });
+
+    it("should throw a InvalidCodeError if a user tries to validate an incorrect code", async () => {
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "authentication-code":
+              return JSON.stringify({ code: "000000" });
+          }
+        },
+      });
+
+      try {
+        await caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' });
+        throw new Error('Should throw')
+      } catch (err: any) {
+        if (err.message !== 'Invalid code') {
+          throw new Error('Should throw InvalidCode')
+        }
+      }
+    });
+
+    it("should throw a AuthenticationCodeExpiredError if a user tries to validate an incorrect code", async () => {
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "authentication-code":
+              return JSON.stringify({ code: "123456", expireAt: 1684757783145 });
+          }
+        },
+      });
+
+      try {
+        await caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' });
+        throw new Error('Should throw')
+      } catch (err: any) {
+        if (err.message !== 'Authentication code expired') {
+          throw new Error('Should throw AuthenticationCodeExpiredError')
+        }
+      }
+    });
+
+    it("should add a new connection to the profile if it does not exist", async () => {
+      const storage: { [key: string]: string } = {};
+
+      const caller = createCaller({
+        get: async (key: string) => {
+          switch (key) {
+            case "profile":
+              return null;
+            case "authentication-code":
+              return JSON.stringify({
+                code: "123456", expireAt: 1784757783145, authParams: {
+                  client_id: "clientId"
+                }
+              });
+          }
+        },
+        put: async (key: string, value: string) => {
+          storage[key] = value;
+        }
+      });
+
+      await caller.validateAuthenticationCode({ code: "123456", email: 'test@example.com', tenantId: 'tenantId' });
+
+      const profile = JSON.parse(storage.profile)
+
+      expect(profile.email).toBe("test@example.com")
+    });
+  });
```

Comment:
would be interesting to try and get Istanbul working so we can do code coverage. I know code coverage isn't the be all and end all *at all* but it helps the person writing the tests to see what they're not testing

Reply:
Seems to be bundled with jest so it's just adding a --coverage flag

Reply:
added a test-with-coverage script now